File: CLDemo-OC/CLDemo-OC/Class/Base/CLTabbarController/Controller/CLTabbarController.m
//
//  TabbarController.m
//  CLDemo
//
//  Created by JmoVxia on 2016/11/17.
//  Copyright © 2016年 JmoVxia. All rights reserved.
//

#import "CLTabbarController.h"
#import "CLNavigationController.h"
#import "CLCollectionController.h"
#import "CLCurriculumController.h"
#import "CLHomepageController.h"
#import "CLCustomTabbar.h"
#import "Tools.h"
#import "CLMyController.h"
#import "UIColor+CLHex.h"
@interface CLTabbarController ()
{
    NSTimeInterval _lastSameIndexTapTime;
    int _tapsInSuccession;
}
@end

@implementation CLTabbarController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self initUI];
    
   }
- (void)initUI {
    NSLog(@"沙盒路径----->>>%@",Tools.pathDocuments);
    
    CLCustomTabbar *tabbar = [[CLCustomTabbar alloc] init];
    __weak __typeof(self) weakSelf = self;
    [tabbar setBulgeCallBack:^{
        __typeof(&*weakSelf) strongSelf = weakSelf;
        strongSelf.selectedIndex = 2;
    }];
    [self setValue:tabbar forKeyPath:@"tabBar"];
    
    self.view.backgroundColor = [UIColor whiteColor];
    [UITabBar appearance].unselectedItemTintColor = [UIColor colorWithHex:@"999999"];
    [UITabBar appearance].tintColor = [UIColor colorWithHex:@"24C065"];
    
    [self addChild:[[CLHomepageController alloc] init]
             title:NSLocalizedString(@"主页", nil)
             image:[UIImage imageNamed:@"tabBar_friendTrends_icon"]
     selectedImage:[UIImage imageNamed:@"tabBar_friendTrends_click_icon"]];
    
    [self addChild:[[CLCurriculumController alloc] init]
             title:NSLocalizedString(@"课程", nil)
             image:[UIImage imageNamed:@"tabBar_new_icon"]
     selectedImage:[UIImage imageNamed:@"tabBar_new_click_icon"]];

    [self addChild:[[CLController alloc] init]
             title:NSLocalizedString(@"课程", nil)
             image:[UIImage imageNamed:@"tabBar_new_icon"]
     selectedImage:[UIImage imageNamed:@"tabBar_new_click_icon"]];

    [self addChild:[[CLCollectionController alloc] init]
             title:NSLocalizedString(@"收藏", nil)
             image:[UIImage imageNamed:@"tabBar_me_icon"]
     selectedImage:[UIImage imageNamed:@"tabBar_me_click_icon"]];

    [self addChild:[[CLController alloc] init]
             title:NSLocalizedString(@"我的", nil)
             image:[UIImage imageNamed:@"tabBar_essence_icon"]
     selectedImage:[UIImage imageNamed:@"tabBar_essence_click_icon"]];
}

-(void)addChild:(UIViewController *)child title:(NSString *)title image:(UIImage *)image selectedImage:(UIImage *)selectedImage {
    child.title = title;
    [child.tabBarItem setTitleTextAttributes:@{NSFontAttributeName: [UIFont systemFontOfSize:11]} forState:UIControlStateNormal];
    [child.tabBarItem setTitleTextAttributes:@{NSFontAttributeName: [UIFont systemFontOfSize:11]} forState:UIControlStateSelected];
    child.tabBarItem.image = [image imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    child.tabBarItem.selectedImage = [selectedImage imageWithRenderingMode:UIImageRenderingModeAlwaysOriginal];
    CLNavigationController *navController = [[CLNavigationController alloc] initWithRootViewController:child];
    [self addChildViewController:navController];
}

-(BOOL)shouldAutorotate {
    if ([self.selectedViewController isKindOfClass:[UINavigationController class]]) {
        return [[(UINavigationController *)self.selectedViewController topViewController] shouldAutorotate] ?: NO;
    } else {
        return [self.selectedViewController shouldAutorotate] ?: NO;
    }
}

-(UIInterfaceOrientationMask)supportedInterfaceOrientations {
    if ([self.selectedViewController isKindOfClass:[UINavigationController class]]) {
        return [[(UINavigationController *)self.selectedViewController topViewController] supportedInterfaceOrientations] ?: UIInterfaceOrientationMaskPortrait;
    } else {
        return [self.selectedViewController supportedInterfaceOrientations] ?: UIInterfaceOrientationMaskPortrait;
    }
}

-(UIInterfaceOrientation)preferredInterfaceOrientationForPresentation {
    if ([self.selectedViewController isKindOfClass:[UINavigationController class]]) {
        return [[(UINavigationController *)self.selectedViewController topViewController] preferredInterfaceOrientationForPresentation] ?: UIInterfaceOrientationPortrait;
    } else {
        return [self.selectedViewController preferredInterfaceOrientationForPresentation] ?: UIInterfaceOrientationPortrait;
    }
}

-(UIStatusBarStyle)preferredStatusBarStyle {
    if ([self.selectedViewController isKindOfClass:[UINavigationController class]]) {
        return [[(UINavigationController *)self.selectedViewController topViewController] preferredStatusBarStyle] ?: UIStatusBarStyleDefault;
    } else {
        return [self.selectedViewController preferredStatusBarStyle] ?: UIStatusBarStyleDefault;
    }
}

-(BOOL)prefersStatusBarHidden {
    if ([self.selectedViewController isKindOfClass:[UINavigationController class]]) {
        return [[(UINavigationController *)self.selectedViewController topViewController] prefersStatusBarHidden] ?: NO;
    } else {
        return [self.selectedViewController prefersStatusBarHidden] ?: NO;
    }
}

-(UIUserInterfaceStyle)overrideUserInterfaceStyle {
    if (@available(iOS 13.0, *)) {
        if ([self.selectedViewController isKindOfClass:[UINavigationController class]]) {
            return [[(UINavigationController *)self.selectedViewController topViewController] overrideUserInterfaceStyle] ?: UIUserInterfaceStyleLight;
        } else {
            return [self.selectedViewController overrideUserInterfaceStyle] ?: UIUserInterfaceStyleLight;
        }
    }else {
        return UIUserInterfaceStyleLight;
    }
}
@end

File: CLDemo-Swift/CLDemo-Swift/Class/Base/CLTabbarController/Controller/CLTabBarController.swift
//
//  CLTabBarController.swift
//  CL
//
//  Created by Chen JmoVxia on 2020/10/16.
//  Copyright © 2020 JmoVxia. All rights reserved.
//

import SnapKit
import UIKit

// MARK: - JmoVxia---类-属性

class CLTabBarController: UITabBarController {
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
    }

    convenience init() {
        self.init(nibName: nil, bundle: nil)
    }

    @available(*, unavailable)
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    deinit {}

    private lazy var tabbar: CLCustomTabbar = {
        let view = CLCustomTabbar()
        view.bulgeCallBack = { [weak self] value in
            guard let self else { return }
            selectedIndex = 2
            tabBar(tabbar, didSelect: value)
        }
        return view
    }()
}

// MARK: - JmoVxia---生命周期

extension CLTabBarController {
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        initUI()
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
    }

    override func viewDidDisappear(_ animated: Bool) {
        super.viewDidDisappear(animated)
    }

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
    }
}

// MARK: - JmoVxia---布局

private extension CLTabBarController {
    func initUI() {
        setValue(tabbar, forKey: "tabBar")
        addChild(CLHomePageController(), title: "首页".localized, image: UIImage(named: "index"), selectedImage: UIImage(named: "indenxHover"))
        addChild(CLChatContactController(), title: "咨询".localized, image: UIImage(named: "advisoty"), selectedImage: UIImage(named: "advisotyHover"))
        addChild(CLRecordController(), title: "记录".localized, image: UIImage(named: "task01"), selectedImage: UIImage(named: "task"))
        addChild(CLMealController(), title: "营养".localized, image: UIImage(named: "meal"), selectedImage: UIImage(named: "mealHover"))
        addChild(CLMeController(), title: "我的".localized, image: UIImage(named: "me"), selectedImage: UIImage(named: "meHover"))
    }
}

// MARK: - JmoVxia---override

extension CLTabBarController {
    override func tabBar(_ tabBar: UITabBar, didSelect item: UITabBarItem) {
        let index = tabBar.items?.firstIndex(of: item) ?? 0
        animationItem(with: index)
    }

    @available(iOS 13.0, *) override var overrideUserInterfaceStyle: UIUserInterfaceStyle {
        set {}
        get {
            if let navigationController = selectedViewController as? UINavigationController {
                navigationController.topViewController?.overrideUserInterfaceStyle ?? .light
            } else {
                selectedViewController?.overrideUserInterfaceStyle ?? .light
            }
        }
    }
}

extension CLTabBarController {
    // 是否支持自动转屏
    override var shouldAutorotate: Bool {
        guard let navigationController = selectedViewController as? UINavigationController else { return selectedViewController?.shouldAutorotate ?? false }
        return navigationController.topViewController?.shouldAutorotate ?? false
    }

    // 支持哪些屏幕方向
    override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        guard let navigationController = selectedViewController as? UINavigationController else { return selectedViewController?.supportedInterfaceOrientations ?? .portrait }
        return navigationController.topViewController?.supportedInterfaceOrientations ?? .portrait
    }

    // 默认的屏幕方向（当前ViewController必须是通过模态出来的UIViewController（模态带导航的无效）方式展现出来的，才会调用这个方法）
    override var preferredInterfaceOrientationForPresentation: UIInterfaceOrientation {
        guard let navigationController = selectedViewController as? UINavigationController else { return selectedViewController?.preferredInterfaceOrientationForPresentation ?? .portrait }
        return navigationController.topViewController?.preferredInterfaceOrientationForPresentation ?? .portrait
    }

    /// 状态栏样式
    override var preferredStatusBarStyle: UIStatusBarStyle {
        guard let navigationController = selectedViewController as? UINavigationController else { return selectedViewController?.preferredStatusBarStyle ?? .default }
        return navigationController.topViewController?.preferredStatusBarStyle ?? .default
    }

    /// 是否隐藏状态栏
    override var prefersStatusBarHidden: Bool {
        guard let navigationController = selectedViewController as? UINavigationController else { return selectedViewController?.prefersStatusBarHidden ?? false }
        return navigationController.topViewController?.prefersStatusBarHidden ?? false
    }
}

// MARK: - JmoVxia---私有方法

private extension CLTabBarController {
    func addChild(_ child: UIViewController, title: String, image: UIImage?, selectedImage: UIImage?) {
        child.title = title
        child.tabBarItem.setTitleTextAttributes([.font: UIFont.boldPingFangSC(11), .foregroundColor: UIColor("333333")], for: .normal)
        child.tabBarItem.setTitleTextAttributes([.font: UIFont.boldPingFangSC(11), .foregroundColor: UIColor("666666")], for: .selected)
        child.tabBarItem.image = image?.withRenderingMode(.alwaysOriginal)
        child.tabBarItem.selectedImage = selectedImage?.withRenderingMode(.alwaysOriginal)
        let navController = CLNavigationController(rootViewController: child)
        addChild(navController)
    }

    func animationItem(with index: Int) {
        if let `class` = NSClassFromString("UITabBarButton") {
            let view = tabBar.subviews.filter { $0.isKind(of: `class`) }[index]
            let animation = CASpringAnimation(keyPath: "transform.scale")
            animation.mass = 0.6
            animation.stiffness = 80
            animation.damping = 10
            animation.initialVelocity = 0.5
            animation.duration = 0.35
            animation.fromValue = 0.25
            animation.toValue = 1
            animation.timingFunction = CAMediaTimingFunction(controlPoints: 0.6, 0.1, 0.30, 0.90)
            view.layer.add(animation, forKey: nil)
        }
    }
}
